content/Refactoring - Sebastian Herrera.md

```diff
@@ -0,0 +1,144 @@
+
+![Code Smells Example](images/code_smells_example.png)
+
+Here you can find a [catalog of code smells](https://github.com/lee-dohm/code-smells/blob/master/README.md), defined mainly by Martin Fowler and Kent Beck.
+
+I highly recommend reading, understanding and memorizing them. If you do, whenever you are doing code review to another person, or just reviewing your own code, you will see much more easily the possible improvements.
+
+**Conclusion: Pay attention to code smells. They might be whispering (or even shouting!) asking for a better solution.**
+
+## 4) ROLE OF TESTING
+By definition, refactoring is the process of changing the code structure, without altering its behavior.
+- How do we know that we have not altered its behavior then?
+  - Yes, you’ve guessed. We need tests.
+- And what if the code we want to refactor doesn’t have enough tests?
+  - Then you should first write the tests, and then refactor.
+
+You may be asking other things like whether it's better to have more unit tests or integration tests, or other things like if it's necessary to have near 100% code coverage on tests.
+
+The short answer is, there is no specific technique or condition to be met in order to be sure that your code works exactly as expected. There is a quote from Dijkstra that says *"Program testing can be used to show the presence of bugs, but never to show their absence."*
+
+So test the way you prefer, and try to cover as much as possible without adding unnecessary duplicated tests. Keep in mind that this will allow you to apply refactoring with much more confidence.
+
+**Conclusion: Having tests is a precondition for refactoring, unless you want to take the risk of breaking things.**
+
+## 5) MECHANICS OF A REFACTORING
+After this long introduction, we are finally reaching the part where I present a couple of examples. Before doing that, I would like to mention that the refactoring as described by Martin Fowler in his book, they have a format, and they have a mechanics.
+
+Format of a refactoring:
+- They have a name
+  - Is important to building a vocabulary of refactoring.
+- They have a motivation
+  - Describes why the refactoring should be done and describes circumstances in which it shouldn’t be done.
+- They have a mechanic
+  - Step-by-step description of how to carry out the refactoring.
+
+Mechanics of a refactoring:
+- Each step in the refactoring is as small as possible.
+- Test after every step. 
+- In practice, if you want to take larger steps, you can, but if the tests fail, then you should go backwards and take smaller steps.
```

Comment:
```suggestion
- In practice, if you want to take larger steps, you can, but if the tests fail, then you should go backward and take smaller steps.
```

Reply:
Ha, seems like `backwards` is used in UK and `backward` in US.